File: BIAN_YUAN_Code/435.java
import java.util.Arrays;
import java.util.Comparator;

/**
 * 
 * 
 * @author tianning.htn
 *
 */
public class Maths {

	public static class Interval {
		int start;
		int end;

		Interval() {
			start = 0;
			end = 0;
		}

		Interval(int s, int e) {
			start = s;
			end = e;
		}
	}
	/**
	 * 适应性归并排序(Adaptive Mergesort)
	 * 
	 */
	static class myComparator implements Comparator<Interval> {
        public int compare(Interval a, Interval b) {
            return a.end - b.end;
        }
    }

	public static int nonOverlappingIntervals(Interval[] intervals) {
		int count = 0;
		if (intervals.length == 0) {
			return 0 ;
		}
		Arrays.sort(intervals , new myComparator());
		Interval current = intervals[0];
		for (int i = 1; i < intervals.length; i++) {
			if (current.end <= intervals[i].start) {
				current = intervals[i];
			}else{
				count++;
			}
		}
		return count;
	}

	public static void main(String[] args) {
		System.out.println(nonOverlappingIntervals(new Interval[]{ new Interval(1,2) , new Interval(1,4),new Interval(2,3),new Interval(2,4)}));
	}
}

File: BIAN_YUAN_Code/435.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e


class Solution(object):
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        
        Sort By the End! And use greedy
        """
        if len(intervals)==1:
            return 0
            
        sortedInt=sorted(intervals , key=lambda x: (x.end, x.start))
        i,j,count=0,1,0
        while j<len(intervals):
            if sortedInt[j].start>=sortedInt[i].end:
                i=j
                j+=1
                
            else:
                count+=1
                j+=1
                
        return count
                
 #---------------------------------Compare to a Top solution-------------------------------------
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        if not intervals: return 0
        intervals.sort(key=lambda x: x.start)  # sort on start time
        currEnd, cnt = intervals[0].end, 0
        for x in intervals[1:]:
            if x.start < currEnd:  # find overlapping interval
                cnt += 1
                currEnd = min(currEnd, x.end)  # erase the one with larger end time
            else:
                currEnd = x.end   # update end time
        return cnt
